x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ApplicationPrivilege.java

```diff
@@ -193,7 +193,7 @@ private static ApplicationPrivilege resolve(String application, Set<String> name
                 if (descriptor != null) {
                     patterns.addAll(descriptor.getActions());
                 } else {
-                    throw new IllegalArgumentException("unknown application privilege [" + name + "]");
+                    patterns.add(name);
```

Comment:
When I changed this to just ignore it, and I created a user with the `kibana_user` role (without inserting the privileges), I was seeing it match unexpected privilege names:

```
PUT {{es}}/_xpack/security/user/kb
Content-Type: application/json
Authorization: Basic elastic changeme

{
  "password" : "password",
  "roles" : [ "kibana_user" ],
  "full_name" : "Admin"
}
```

```
POST {{es}}/_xpack/security/user/_has_privileges
Content-Type: application/json
Authorization: Basic kb password

{
    "applications":[
        {
            "application":"kibana-.kibana",
            "resources":["*"],
            "privileges":[
              "read",
              "action:foo",
              "all",
              "i-dont-exist"
            ]
        }
    ]
}
```

```
HTTP/1.1 200 OK
content-type: application/json; charset=UTF-8
content-encoding: gzip
content-length: 144

{
  "username": "kb",
  "has_all_requested": false,
  "cluster": {},
  "index": {},
  "application": {
    "kibana-.kibana": {
      "*": {
        "all": true,
        "read": true,
        "action:foo": false,
        "i-dont-exist": true
      }
    }
  }
}
```

If you have any available bandwidth, any assistance would be much appreciated.

Reply:
I think we just need soemthing like:

```
if (patterns.isEmpty()) {
  return NONE.apply(application);
} 
```
We didn't have it before because `get` has a `names.isEmpty` check, and it wasn't previously possible for non-empty names to produce non-empty patterns.

Or maybe it was, if a the named privileges had no actions in them. That sounds like it was probably always a bug...

Reply:
OK, it turns out that `NONE` doesn't work (in general) - I'll sort out a fix.

Reply:
I'm tired and my brain hurts, so I'm going to have to leave this for now.
The fundamental problem is that a Privilege with no actions, implies another privilege with no actions.
So if "all" doesn't exist, and "read" doesn't exist, and we treat them as having actions, then they match each other.

That's logical, but not what we want. I think we want to say that a privilege with no actions matches nothing, except up to this point we've said that actions are optional, which clearly wasn't' working, so I _think_ we either need to
1. say actions are required
2. the privilege name is an implicit action _all the time_ (so "abc" implies "abc", but not "xyz").

I'm not making a decision on that at 6pm on a Friday, but my inclination is to go with (1).

Reply:
I realised option (2) doesn't work. If the privilege name is always an implicit action, then "all" would not imply "read".

The only way something like that would work is we added the privilege name if it had no actions, but in that case we could just require that the person defining the privileges did that explicitly rather than implicitly.

So, I'm going to make actions required, and if some app doesn't need them, then they just give everything a single action of `action:<privielge-name>`.

Reply:
I opened a PR for this - https://github.com/elastic/elasticsearch/pull/32272

I believe the consequence for this PR is two-fold:
1. You can stop adding the name to the patterns list if the privilege doesn't exist
2. If the privileges have not been defined yet, then "read" will _not_ imply "read" because both are meaningless references to non-existing privileges. So `_has_privileges` will return false for all Kibana privileges until such a time as they have been stored through the API.

I feel like that's the correct response. Treating privileges with the same name as special, even though they haven't been defined is messy. It means "all" implies "all", and "read" implies "read", but "all" does not imply "read". A blanket "these don't exist", so they can't be true is a safer security position.

I would be happy to go back to the previous state where calling _has_privileges with arguments that aren't defined is an error, but I don't want to introduce leniency around special matching of names that aren't defined.

Reply:
This is great, thanks so much @tvernum!